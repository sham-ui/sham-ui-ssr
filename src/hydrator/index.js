import Storage from './storage/service';
import ServerSideHooks from './hooks/service';
import ServerSideDOM from './dom/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Main function for setup hydrating components
 * @param {Object} DI App DI container
 */
export function setup( DI ) {
    const storage = new Storage();

    // Override default hooks service
    DI.bind(
        HOOKS_DI_KEY,
        new ServerSideHooks( storage )
    );

    // Override default dom service
    DI.bind(
        DOM_DI_KEY,
        new ServerSideDOM( storage )
    );

    return storage.root;
}


/**
 * Hydrate all components
 * @param {Object} DI App DI container
 * @return {Promise}
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import { setupHydrator, hydrate } from 'sham-ui-ssr/hydrator';
 *
 * const DI = createDI();
 * const root = setupHydrator( DI );
 * mainInitializer( DI, root );
 * start( DI );
 * hydrate( DI ).then( storage => {
 *     const { html, data } = storage.hydrate();
 * } )
 */
export function hydrate( DI ) {
    return DI.resolve( HOOKS_DI_KEY ).hydrating();
}
