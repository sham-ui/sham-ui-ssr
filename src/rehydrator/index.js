import RestoreDom from './dom/service';
import RehydrateHooks from './hooks/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Save hooks & dom services for restore after rehydrating
 * @param {Object} DI App DI container
 * @return {Function}
 */
function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

/**
 * Main function for rehydrating
 * @param {Object} DI App DI container
 * @param {Object} storage
 * @return {Function} Restore function for disable rehydrating and switch to default behavior
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import setupUnsafe from 'sham-ui-unsafe';
 * import { setupRehydrator } from 'sham-ui-ssr/rehydrator';
 *
 * const DI = createDI();
 * const data = JSON.parse( document.getElementById( 'data' ).innerText );
 * const disableRehydrating = setupRehydrator( DI, data );
 * mainInitializer( DI, root );
 * start( DI );
 * disableRehydrating();
 * setupUnsafe( DI );
 */
export function setup( DI, storage ) {
    const restore = makeRestore( DI );
    DI.bind( HOOKS_DI_KEY, new RehydrateHooks( storage ) );
    DI.bind( DOM_DI_KEY, new RestoreDom( storage ) );
    return restore;
}
