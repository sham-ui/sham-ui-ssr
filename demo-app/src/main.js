import './styles/main.scss';
import setupUnsafe from 'sham-ui-unsafe';
import { createDI, start } from 'sham-ui';
import { setup as setupHydrator, hydrate } from '../../src/hydrator/index';
import { setup as setupRehydrator } from '../../src/rehydrator/index';
import mainInitializer from './initializers/main';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

const DI = createDI();
const restoreOriginal = makeRestore( DI );

const root = setupHydrator( DI );
mainInitializer( DI, root );
start( DI );
hydrate( DI ).then(
    ( storage ) => {
        const getById = document.getElementById.bind( document );
        // eslint-disable-next-line no-console
        console.dir( storage );
        try {
            const { html, data } = storage.hydrate();
            getById( 'ssr-rendered' ).innerHTML = html;
            getById( 'raw-html' ).innerText = html;
            getById( 'data' ).innerText = JSON.stringify( JSON.parse( data ), null, 4 );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
        try {
            DI.resolve( 'sham-ui:store' ).byId.clear();
            const root = getById( 'client' );
            root.innerHTML = getById( 'ssr-rendered' ).innerHTML;
            restoreOriginal();
            const data = JSON.parse( document.getElementById( 'data' ).innerText );
            const disableRehydrating = setupRehydrator( DI, data );
            mainInitializer( DI, root );
            start( DI );
            disableRehydrating();
            setupUnsafe( DI );
            // eslint-disable-next-line no-console
            console.dir( DI.resolve( 'sham-ui:store' ).byId.keys() );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
    }
);

