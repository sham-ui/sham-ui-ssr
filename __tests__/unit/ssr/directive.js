import { compile } from 'sham-ui-test-helpers';
import { ssr } from '../../../src/testing';

beforeEach( () => {
    window.Custom = compile`
        <div id="custom" :ref={{name}}>{{name}}</div>
    `;
    window.Ref = class {
        constructor( component ) {
            this.component = component;
            this.node = null;
            this.name = null;
        }

        bind( node ) {
            this.node = node;
        }

        unbind() {
            this.node = null;
            if ( this.name ) {
                delete this.component[ this.name ];
            }
            this.name = null;
        }

        update( name ) {
            this.name = name;
            this.component[ name ] = this.node;
        }
    };
} );

afterEach( () => {
    delete window.Custom;
    delete window.Ref;
} );

it( 'should be trimmed from html', async() =>{
    expect.assertions( 2 );

    function directiveMock() {
        return function() {
            this.bind = function() {};
            this.update = function() {};
            this.unbind = function() {};
        };
    }

    const directives = {
        fade: directiveMock(),
        show: directiveMock(),
        content: directiveMock()
    };

    const meta = await ssr(
        compile`
            <div :fade/>
            <div :show={{ visible }}/>
            <div :content="come {{ text }}"/>
        `,
        {
            directives
        }
    );
    expect( meta.html ).toBe( '<div></div><div></div><div></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'methods bind, update, unbind should be called', async() => {
    expect.assertions( 4 );

    const directive = jest.fn();
    directive.prototype.bind = jest.fn();
    directive.prototype.update = jest.fn();

    const directives = { directive };

    const meta = await ssr(
        compile`<div :directive={{ value }}/>`,
        {
            directives,
            value: true
        }
    );
    expect( directive ).toHaveBeenCalledWith( meta.component );
    expect( directive.prototype.bind ).toHaveBeenCalledWith( meta.component.nodes[ 0 ] );
    expect( directive.prototype.update ).toHaveBeenCalledWith( true );

    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'ref directive', async() => {
    expect.assertions( 1 );

    const meta = await ssr(
        compile`
            <div>
                <div id="foo" :ref="foo">
                    {% if test %}
                        <div id="test-true" :ref={{ test + "Inner"}}></div>
                    {% else %}
                        <div id="test-false" :ref={{ test + "Inner"}}></div>
                    {% endif %}
                </div>            
            </div>
        `,
        {
            directives: {
                ref: window.Ref
            },
            test: true
        }
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'ref directive with custom tag', async() => {
    expect.assertions( 2 );

    const meta = await ssr(
        compile`
            <div>
                <div id="foo" :ref="foo">
                    {% if test %}
                        <Custom name="foo"/>
                    {% else %}
                        <Custom name="bar"/>
                    {% endif %}
                </div>            
            </div>
        `,
        {
            directives: {
                ref: window.Ref
            },
            test: true
        }
    );
    expect( meta.html ).toBe(
        '<div><div id="foo"><div id="custom">foo</div><!--0--></div></div>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

