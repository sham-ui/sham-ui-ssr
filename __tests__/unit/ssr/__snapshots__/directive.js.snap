// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`methods bind, update, unbind should be called 1`] = `
Object {
  "data": "{
    \\"component\\": {
        \\"N\\": [
            0
        ]
    }
}",
  "html": "<div></div>",
}
`;

exports[`ref directive 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"N\\": [
            0
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                0
            ]
        ]
    }
}",
  "html": "
<div>
  <div id=\\"foo\\">
    <div id=\\"test-true\\"></div>
  </div>
</div>
",
}
`;

exports[`ref directive with custom tag 1`] = `
Object {
  "data": "{
    \\"1\\": {
        \\"C\\": [
            \\"2\\"
        ],
        \\"N\\": [
            1
        ]
    },
    \\"2\\": {
        \\"N\\": [
            0,
            [
                0,
                0
            ]
        ]
    },
    \\"component\\": {
        \\"C\\": [
            \\"1\\"
        ],
        \\"N\\": [
            0,
            [
                0,
                0
            ]
        ]
    }
}",
  "html": "
<div>
  <div id=\\"foo\\">
    <div id=\\"custom\\">foo</div>
    <!--0-->
  </div>
</div>
",
}
`;

exports[`should be trimmed from html 1`] = `
Object {
  "data": "{
    \\"component\\": {
        \\"N\\": [
            0,
            1,
            2
        ]
    }
}",
  "html": "
<div></div>
<div></div>
<div></div>
",
}
`;
